### PR TITLE
Event aggregation and rewards tracking

### LINKED ISSUE #298: Feature:  fetch events by block
There is a request to get/watch for Arc-related events as each block is generated, returning the found set of events in the order in which they were fired in each block.

### LINKED ISSUE #299: Create event that merges CR and GP redeem events
Create an event that merges CR and GP redeem events into a single PubSub event.

### PR DESCRIPTION
**Pull request by dkent600**
Resolves #299 
Starts on  #298 

- adds `Redeemer.rewardsEvents`  that enables you to get, watch and subscribe to events that return a `RedeemerRewardEventsResult` when rewards are rewarded, either via `Redeemer.redeem` or directly via `GenesisProtocol` and `ContributionReward`

- adds `Web3EventService.aggregatedEventsFetcherFactory` that, given a list of contract events, returns an `EntityFetcherFactory` that enables you to get, watch and subscribe to an event that fires whenever a transaction is detected that emitted one or more of the given contract events.

**Note* these functions do not allow you to supply a web3Filter specifying from and to blocks.  They always get the latest "pending" transactions and wait for them to be mined to an optionally-given depth.  The did not see past information, only going forward.

**Breaking**
- removed `suppressDups` option from the webFilter (`EventFetcherFilterObject`) for event fetchers.

### CODE REVIEW

**Comment by dev-matan-tsuberi** [comment]
Please don't write "Begins to resolve..." because github will close that issue.

**Comment by tibetsprague** [comment]
I dont think this will be very useful if we cant pass from and to blocks in

**Comment by dkent600** [comment]
@tibetsprague 

> I dont think this will be very useful if we cant pass from and to blocks in

What is your specific use case?

File: lib/web3EventService.ts [hunk 1]
```diff
@@ -296,27 +213,289 @@ export class Web3EventService {
...
+  }
+
+  /**
+   * Convert the EntityFetcherFactory<TEntitySrc, TEntityOriginalSrc> into an
+   * EntityFetcherFactory<TEntityDest, TEntitySrc>.
+   *
+   * @param entityFetcherFactory The source EntityFetcherFactory
+   * @param transformEventCallback Converts TEntitySrc into TEntityDest
+   */
+  public pipeEntityFetcherFactory<TEntityDest, TEntitySrc, TEntityOriginalSrc>(
```
Comment (tibetsprague) [inline lib/web3EventService.ts]: What does T stand for at the beginning of the types?

**Comment by tibetsprague** [comment]
the way the cache script works is that we load the last block cached, then find the latest block and then look for all events that happened from the last one to the latest one. We load the latest one first and use that as the to block to make sure that all the events we find are only up to the latest block and we dont find some in one latest block and some others later in the caching process from a different latest block. that way we can always be sure we have all the events up to a particular block and no more. make sense? So we need to be able to specify from and to.

**Comment by dkent600** [comment]
@tibetsprague 

> I gotta say, its very hard for me to review your code because it is so dense and full of so many types, generic types, interfaces, factories, fetchers... I cant tell how much this is mostly because of the nature of the problems you are solving or if there is some unnecessary over engineering going on

The lint rules require types on almost everything, and though sometimes it is a pain, it is generally worth it for catching errors.

The Web3EventService code is complicated and I think would benefit from reorganizing for clarity and maybe a little refactoring.

File: lib/web3EventService.ts (hunk 1 above)
Comment (dkent600) [reply lib/web3EventService.ts]: `T` stands for `Type`

File: lib/web3EventService.ts [hunk 2]
```diff
@@ -296,27 +213,289 @@ export class Web3EventService {
...
+        const txReceipt = await TransactionService.watchForMinedTransaction(
+          txHash,
+          undefined,
+          requiredDepth) as TransactionReceipt;
+
+        for (const log of txReceipt.logs) {
+
+          const contractAddress = log.address;
+          /**
+           * see if a contract of interest generated this log
```
[the remark this answers is not in the document]
Comment (dkent600) [reply lib/web3EventService.ts]: I usually use comments to introduce what I'm doing in the following line(s).  That is what this comment is doing.

File: lib/web3EventService.ts [hunk 3]
```diff
@@ -296,27 +213,289 @@ export class Web3EventService {
...
+   * This only finds transactions that happen after you have initiated the get or watch.  It does
+   * not look back in time.
+   *
+   * The returned EntityFetcher ignores the `argFilter` (irrelevant) and `web3Filter` (always "pending") arguments.
+   * The `get`, `watch` and `subscribe` methods ignore the `requiredDepth` parameter, which you must pass here.
+   *
+   * @param events An array of `EventToAggregate` that specifies which events to look for,
+   * by name and Arc.js contract wrapper.
+   * @param requiredDepth -- If set then will not invoke the callback until the transaction has been mined to
+   * the requiredDepth.Pass -1 to use the Arc.js's global default depth.  Note that the Fetcher
```
[the remark this answers is not in the document]
Comment (dkent600) [reply lib/web3EventService.ts]: Fixed

**Comment by dkent600** [comment]
There is no immediate need for this as is, and it needs to be enabled to work on block number ranges.

**Pull request closed without merging**